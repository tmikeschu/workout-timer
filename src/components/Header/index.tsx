import * as React from "react";
import { styled, ThemeSwitch } from "theme";
import { toHoursMinutesSeconds, fromHoursMinutesSeconds } from "utils";
import { AppMachineContext } from "contexts/machine";

const Container = styled.div`
  padding: 1rem;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  h2 {
    width: 100%;
  }
`;

const Time = styled.h2`
  font-size: 5rem;
  text-align: center;
  margin: 1rem 0;
  display: flex;
  justify-content: center;
`;

const Header: React.FC<{}> = () => {
  const [current, send] = React.useContext(AppMachineContext);

  const hms = toHoursMinutesSeconds(current.context.currentTime);

  const changeHours = (e: React.FocusEvent<HTMLInputElement>): void => {
    const hours = Number(e.target.innerText);
    const time = fromHoursMinutesSeconds([hours, hms[1], hms[2]]);
    send({ type: "SET_TIME", payload: { time } });
  };

  const changeMinutes = (e: React.FocusEvent<HTMLInputElement>): void => {
    const minutes = Number(e.target.innerText);
    const time = fromHoursMinutesSeconds([hms[0], minutes, hms[2]]);
    send({ type: "SET_TIME", payload: { time } });
  };

  const changeSeconds = (e: React.FocusEvent<HTMLInputElement>): void => {
    const seconds = Number(e.target.innerText);
    const time = fromHoursMinutesSeconds([hms[0], hms[1], seconds]);
    send({ type: "SET_TIME", payload: { time } });
  };

  const [hours, minutes, seconds] = hms.map(time =>
    String(time).padStart(2, "0")
  );

  return (
    <Container>
      <h1>Workout Timer</h1>
      <ThemeSwitch />
      <Time>
        <div
          contentEditable={true}
          onBlur={changeHours}
          dangerouslySetInnerHTML={{ __html: hours }}
        />
        :
        <div
          contentEditable={true}
          onBlur={changeMinutes}
          dangerouslySetInnerHTML={{ __html: minutes }}
        />
        :
        <div
          contentEditable={true}
          onBlur={changeSeconds}
          dangerouslySetInnerHTML={{ __html: seconds }}
        />
      </Time>
    </Container>
  );
};

export default Header;
