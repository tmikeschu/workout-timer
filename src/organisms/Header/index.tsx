import * as React from "react";
import { styled, ThemeSwitch } from "theme";
import { toHoursMinutesSeconds, fromHoursMinutesSeconds } from "utils";
import { useAppMachine } from "contexts/machine";
import NumberInput from "atoms/NumberInput";

const Container = styled.div`
  padding: 1rem;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
`;

const Time = styled.div`
  width: 100%;
  font-size: 2rem;
  select {
    font-size: 4rem;
  }
  text-align: center;
  margin: 1rem 0;
  display: flex;
  justify-content: center;
  align-items: center;
  & > input {
    padding: 0;
    max-width: 25%;
  }
`;

const Header: React.FC = () => {
  const [current, send] = useAppMachine();

  const hms = toHoursMinutesSeconds(current.context.currentTime);

  const changeHours = (e: React.ChangeEvent<HTMLSelectElement>): void => {
    const hours = Number(e.target.value);
    const time = fromHoursMinutesSeconds([hours, hms[1], hms[2]]);
    send({ type: "SET_TIME", payload: { time } });
  };

  const changeMinutes = (e: React.ChangeEvent<HTMLSelectElement>): void => {
    const minutes = Number(e.target.value);
    const time = fromHoursMinutesSeconds([hms[0], minutes, hms[2]]);
    send({ type: "SET_TIME", payload: { time } });
  };

  const changeSeconds = (e: React.ChangeEvent<HTMLSelectElement>): void => {
    const seconds = Number(e.target.value);
    const time = fromHoursMinutesSeconds([hms[0], hms[1], seconds]);
    console.log(time);
    send({ type: "SET_TIME", payload: { time } });
  };

  const [hours, minutes, seconds] = hms;

  return (
    <Container data-testid="Header">
      <h1>Workout Timer</h1>
      <ThemeSwitch />
      <Time>
        <NumberInput
          range={[0, 24]}
          data-testid="Header__hours"
          onChange={changeHours}
          value={hours}
          display={(n: number): string => String(n).padStart(2, "0")}
        />
        <span>:</span>
        <NumberInput
          range={[0, 59]}
          data-testid="Header__minutes"
          onChange={changeMinutes}
          value={minutes}
          display={(n: number): string => String(n).padStart(2, "0")}
        />
        <span>:</span>
        <NumberInput
          range={[0, 59]}
          data-testid="Header__seconds"
          onChange={changeSeconds}
          value={seconds}
          display={(n: number): string => String(n).padStart(2, "0")}
        />
      </Time>
    </Container>
  );
};

export default Header;
