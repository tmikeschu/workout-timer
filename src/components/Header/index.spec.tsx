import React from "react";
import { render } from "@testing-library/react";
import Header from "./";

describe("snapshot", () => {
  it("is correct", () => {
    const tree = render(<Header />);
    expect(tree).toMatchSnapshot();
  });
});
