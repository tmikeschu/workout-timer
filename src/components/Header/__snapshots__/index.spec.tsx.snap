// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot is correct 1`] = `
<body>
  <div>
    <div
      class="css-uh9obs"
    >
      <h1>
        Workout Timer
      </h1>
      <button
        class="css-yh1oe3"
      >
        <svg
          height="24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 7a5 5 0 110 10 5 5 0 010-10m0 2a3 3 0 100 6 3 3 0 000-6m0-7l2 3a7 7 0 00-4 0l2-3M3 7h5a7 7 0 00-3 4L3 7m0 10l2-4a7 7 0 003 4H3M21 7l-2 4a7 7 0 00-2-4h4m0 10h-4a7 7 0 002-4l2 4m-9 5l-2-3h4l-2 3z"
          />
        </svg>
      </button>
      <h2
        class="css-10xeong"
      >
        <div
          contenteditable="true"
        >
          00
        </div>
        :
        <div
          contenteditable="true"
        >
          06
        </div>
        :
        <div
          contenteditable="true"
        >
          00
        </div>
      </h2>
    </div>
  </div>
</body>
`;
